File: 1406.石子游戏-iii.py
#
# @lc app=leetcode.cn id=1406 lang=python3
#
# [1406] 石子游戏 III
#

# @lc code=start
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp0, dp1, dp2 = 0, stoneValue[n - 1], 0
        sum = stoneValue[n - 1]
        if n >= 2:
            sum += stoneValue[n - 2]
            dp0 = sum - min(dp1, dp2)
        else:
            dp0 = dp1
        for i in range(n - 3, -1, -1):
            sum += stoneValue[i]
            temp = dp0
            dp0 = sum - min(dp0, dp1, dp2)
            dp2 = dp1
            dp1 = temp
        if dp0 * 2 > sum:
            return 'Alice'
        elif dp0 * 2 < sum:
            return 'Bob'
        else:
            return 'Tie'

# @lc code=end

File: 1406.石子游戏-iii.java
import java.util.Arrays;


/*
 * @lc app=leetcode.cn id=1406 lang=java
 *
 * [1406] 石子游戏 III
 */

// @lc code=start
class Solution {
    public String stoneGameIII(int[] stoneValue) {
        int n = stoneValue.length;
        int[] dp = new int[n + 1];
        dp[n - 1] = stoneValue[n - 1];
        int sum = dp[n - 1];
        if (n - 2 >= 0) {
            sum += stoneValue[n - 2];
            dp[n - 2] = sum - Math.min(dp[n], dp[n - 1]);
        }
        for (int i = n - 3; i >= 0; i--) {
            sum += stoneValue[i];
            dp[i] = sum - Math.min(dp[i + 1], Math.min(dp[i + 2], dp[i + 3]));
        }
        if (dp[0] * 2 > sum) {
            return "Alice";
        } else if (dp[0] * 2 == sum) {
            return "Tie";
        } else {
            return "Bob";
        }
    }
}
// @lc code=end

